src/cpp/core/FileLock.cpp

```diff
@@ -15,163 +15,144 @@
-   #undef BOOST_USE_WINDOWS_H
-#endif
-#include <boost/interprocess/sync/file_lock.hpp>
+#include <boost/algorithm/string.hpp>
 
 namespace rstudio {
 namespace core {
 
-bool FileLock::isLocked(const FilePath& lockFilePath)
-{
-   using namespace boost::interprocess;
+namespace locks {
 
-   // if the lock file doesn't exist then it's not locked
-   if (!lockFilePath.exists())
-      return false;
+namespace {
 
-   // check if it is locked
-   try
-   {
-      file_lock lock(string_utils::utf8ToSystem(lockFilePath.absolutePath()).c_str());
+const char * const kLocksConfPath = "/etc/rstudio-locks.conf";
+
+} // end anonymous namespace
 
-      if (lock.try_lock())
+void initialize()
+{
+   using namespace boost::algorithm;
+   
+   FilePath locksConfPath(kLocksConfPath);
+   if (!locksConfPath.exists())
+      return;
+   
+   std::map<std::string, std::string> conf;
+   Error error = core::readStringMapFromFile(locksConfPath, &conf);
+   if (error)
+      LOG_ERROR(error);
+   
+   if (conf.count("lock-type"))
+   {
+      std::string lockType = to_lower_copy(conf["lock-type"]);
+      if (lockType == "advisory")
       {
-         lock.unlock();
-         return false;
+         FileLock::setLockType(FileLock::LOCKTYPE_ADVISORY);
+      }
+      else if (lockType == "link-based")
+      {
+         FileLock::setLockType(FileLock::LOCKTYPE_LINKBASED);
       }
       else
       {
-         return true;
+         std::string msg = std::string() +
+               "unrecognized lock type '" + lockType + "'";
+         LOG_ERROR_MESSAGE(msg);
       }
    }
-   catch(boost::interprocess::interprocess_exception& e)
+   
+   if (conf.count("lock-refresh-rate"))
    {
-      Error error(boost::interprocess::ec_from_exception(e), ERROR_LOCATION);
-      error.addProperty("lock-file", lockFilePath);
-      LOG_ERROR(error);
-      return false;
+      long refreshRate = ::atol(conf["lock-refresh-rate"].c_str());
+      if (refreshRate == 0)
+         LOG_ERROR_MESSAGE("invalid lock refresh rate '" + conf["lock-refresh-rate"] + "'");
+      else
+         FileLock::setLockRefreshRate(boost::posix_time::seconds(refreshRate));
+   }
+   
+   if (conf.count("lock-timeout-interval"))
+   {
+      long timeoutInterval = ::atol(conf["lock-timeout-interval"].c_str());
+      if (timeoutInterval == 0)
+         LOG_ERROR_MESSAGE("invalid lock timeout interval '" + conf["lock-timeout-interval"] + "'");
+      else
+         FileLock::setLockTimeoutInterval(boost::posix_time::seconds(timeoutInterval));
    }
 }
 
+} // namespace locks
 
-struct FileLock::Impl
-{
-   FilePath lockFilePath;
-   boost::interprocess::file_lock lock;
-};
+// default values for static members
+FileLock::LockType FileLock::type_(FileLock::LOCKTYPE_ADVISORY);
+boost::posix_time::seconds FileLock::timeoutInterval_(30);
+boost::posix_time::seconds FileLock::refreshRate_(20);
 
-FileLock::FileLock()
-   : pImpl_(new Impl())
+boost::shared_ptr<FileLock> FileLock::create()
```

Comment:
For the scenario of the Desktop vs. Server you'll also want to be able to explicitly create an AdvisoryLock (i.e. sometimes you'll ask for Advisory and sometimes you'll ask for the "default")